src/resources/tcp_esp32.cc

```diff
@@ -408,26 +384,44 @@ PRIMITIVE(read)  {
       return Smi::from(-1);
     }
 
-    ByteArray* array = allocate_read_buffer(process, p);
+    if (p->tot_len == 0) return Smi::from(-1);  // Nothing to read.
+
+    int offset = socket->read_offset();
+    // Wifi MTU is 1500 bytes, subtract a 20 byte TCP header and we have 1480.
+    const int MAX_SIZE = 496;  // Three nicely-packable byte arrays per 1480 MTU.
+    int tot_len = p->tot_len;
+    ByteArray* array = process->allocate_byte_array(Utils::min(MAX_SIZE, tot_len - offset), false);  // On-heap byte array.
     if (array == null) ALLOCATION_FAILED;
 
     ByteArray::Bytes bytes(array);
 
-    int offset = 0;
-    while (offset < bytes.length()) {
-      memcpy(bytes.address() + offset, p->payload, p->len);
-      offset += p->len;
-
-      pbuf* n = p->next;
-      // Free the first part of the chain.
-      if (n != null) pbuf_ref(n);
-      pbuf_free(p);
-      p = n;
+    int bytes_to_ack = 0;
+    int copied = 0;
+    while (copied < bytes.length()) {
+      int to_copy = Utils::min(p->len - offset, bytes.length() - copied);
+      uint8* payload = unvoid_cast<uint8*>(p->payload);
+      memcpy(bytes.address() + copied, payload + offset, to_copy);
+      copied += to_copy;
+      offset += to_copy;
+      if (offset == p->len) {
+        pbuf* n = p->next;
+        bytes_to_ack += p->len;
+        // Free the first part of the chain.
+        if (n != null) pbuf_ref(n);
+        pbuf_free(p);
+        p = n;
+        offset = 0;
+      }
     }
 
     socket->set_read_buffer(p);
+    socket->set_read_offset(offset);
```

Comment:
Would it make sense to join these two calls?

```suggestion
    socket->set_read_buffer(p, offset);
```